File: krypto.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

const char PLUS = '+';
const char MINUS = '-';
const char DIVIDE = '/';
const char MULTIPLY = '*';

string formatAnswer(vector<short> cards, string operations)
{
    string answer = "((";

    for (short i = 0; i < 4; i++)
    {
        answer += to_string(cards[i]);

        if (i >= 1 && i < 3)
            answer += ')';

        if (i < 3)
            answer += ' ';

        answer += operations[i];
        answer += ' ';
    }

    return answer;
}

void computeAllOperations(vector<string> &result, string operations = "")
{
    if (operations.size() == 3)
    {
        result.push_back(operations);
        return;
    }

    operations.push_back(PLUS);
    computeAllOperations(result, operations);

    *operations.rbegin() = MINUS;
    computeAllOperations(result, operations);

    *operations.rbegin() = MULTIPLY;
    computeAllOperations(result, operations);

    *operations.rbegin() = DIVIDE;
    computeAllOperations(result, operations);
}

int calculate(vector<short> cards, string operations)
{
    int result = cards[0];

    for (short i = 1; i < 4; i++)
    {
        char operation = operations[i - 1];
        int card = cards[i];

        if (operation == DIVIDE)
        {
            if (card == 0 || result % card != 0)
                return -1;

            result /= card;
        }

        if (operation == MULTIPLY)
            result *= card;
        else if (operation == PLUS)
            result += card;
        else if (operation == MINUS)
            result -= card;
    }

    return result;
}

vector<string> allOperations;
string searchSolution(vector<short> &cards, int target)
{
    if (allOperations.empty())
        computeAllOperations(allOperations);

    sort(cards.begin(), cards.end());

    do
    {
        for (string currentOperations : allOperations)
            if (calculate(cards, currentOperations) == target)
                return formatAnswer(cards, currentOperations);

    } while (next_permutation(cards.begin(), cards.end()));

    return "";
}

int main()
{
    vector<short> cards(4);
    int target;

    for (short &card : cards)
        cin >> card;

    cin >> target;

    string result = searchSolution(cards, target);

    if (result == "")
    {
        cout << "No solution\n";
        return 0;
    }

    cout << result << "= " << target << '\n';
}

File: krypto.py
import itertools, re, sys

PLUS = "+"
MINUS = "-"
DIVIDE = "/"
MULTIPLY = "*"

COMBINATIONS = [
    "((n_n)_n)_n",
    "n_((n_n)_n)",
    "(n_(n_n))_n",
    "n_(n_(n_n))",
    "(n_n)_(n_n)",
]

show_all = False


def format_answer(expression):
    answer = ""

    for character in expression:
        if character in "+-*/":
            answer += " " + character + " "
        else:
            answer += character

    return answer


def compute_all_operations(result, operations=""):
    if len(operations) == 3:
        result.append(operations)
        return

    operations += PLUS
    compute_all_operations(result, operations)

    operations = operations[:-1] + MINUS
    compute_all_operations(result, operations)

    operations = operations[:-1] + MULTIPLY
    compute_all_operations(result, operations)

    operations = operations[:-1] + DIVIDE
    compute_all_operations(result, operations)


def compute_all_combinations(result):
    all_operations = []
    compute_all_operations(all_operations)

    for current_operations in all_operations:
        for combination in COMBINATIONS:
            for operation in current_operations:
                combination = combination.replace("_", operation, 1)

            result.add(combination)


def format_expression(cards, expression, operations):

    for operation in operations:
        expression = expression.replace("_", operation, 1)

    for card in cards:
        expression = expression.replace("n", str(card), 1)

    return expression


def divide(a, b):
    if b == 0:
        raise ZeroDivisionError

    if a % b == 0:
        return a // b

    raise ValueError("Division is not an integer")


def replace_division_expression(expr):
    def replacement(match):
        num1 = match.group(1)
        num2 = match.group(2)
        return f"divide({num1}, {num2})"

    pattern = r"(\d+)/(\d+)"
    return re.sub(pattern, replacement, expr)


def search_solution(cards, target):
    all_combinations, solutions = set(), set()

    compute_all_combinations(all_combinations)

    cards.sort()

    for permuted_cards in itertools.permutations(cards):
        for combination in all_combinations:

            for card in permuted_cards:
                combination = combination.replace("n", str(card), 1)

            try:
                if eval(replace_division_expression(combination)) == target:
                    if not show_all:
                        return [format_answer(combination)]

                    solutions.add(format_answer(combination))
            except ZeroDivisionError:
                pass
            except ValueError:
                pass

    return solutions


def main():
    global show_all
    show_all = "--all" in sys.argv

    cards = [0] * 4
    target = 0

    numbers = input().split()

    for i in range(4):
        cards[i] = int(numbers[i])

    target = int(numbers[4])

    solution = search_solution(cards, target)

    if not solution:
        print("No solution")
        return

    if not show_all:
        print(solution[0] + " = " + str(target))
        return

    for solution in solution:
        print(solution + " = " + str(target))


if __name__ == "__main__":
    main()
